### PR TITLE
Sync x-www-browser with defaultbrowser when using Palemoon petbuild

### PR DESCRIPTION
**Pull request by lakshayrohila**
Use `update-alternatives` command to sync by default, `/usr/bin/x-www-browser` with `/usr/local/bin/defaultbrowser`.

Specifically, this is required in `epdfview` since it uses `x-www-browser` by default to open URL links.

@dimkr, @peabee please review.

### CODE REVIEW

**Comment by dimkr** [comment]
In Vanilla Dpup I went for a different solution, defaultbrowser runs x-www-browser. You set the default browser the Debian way, not the Puppy way.

**Comment by lakshayrohila** [comment]
> In Vanilla Dpup I went for a different solution, defaultbrowser runs x-www-browser. You set the default browser the Debian way, not the Puppy way.

@dimkr, thanks for the suggestion! I have re-implemented this fix to your suggestion. Please re-review.

**Comment by dimkr** [comment]
I think it would be cleaner to change the palemoon petbuild, so it makes x-www-browser run palemoon, then sets defaultbrowser to x-www-browser, only in Debian or Ubuntu based build.

https://github.com/puppylinux-woof-CE/woof-CE/blob/4c8a372cd9bec68d687d2f8018e7121d5563efca/woof-code/packages-templates/firefox_FIXUPHACK#L87

**Comment by lakshayrohila** [comment]
> I think it would be cleaner to change the palemoon petbuild, so it makes x-www-browser run palemoon, then sets defaultbrowser to x-www-browser, only in Debian or Ubuntu based build.
> 
> https://github.com/puppylinux-woof-CE/woof-CE/blob/4c8a372cd9bec68d687d2f8018e7121d5563efca/woof-code/packages-templates/firefox_FIXUPHACK#L87

@dimkr, done, please review.

---

One question: is `pinstall.sh` run in `chmod`? Reading some lines of code in `3builddistro`, I don't think it does. Then I'll have to update this commit.

File: woof-code/rootfs-petbuilds/palemoon/pinstall.sh
Comment (dimkr) [inline woof-code/rootfs-petbuilds/palemoon/pinstall.sh]: pinstall.sh runs without `chroot` during woof-CE, you must use relative paths - usr/local/bin/defaulbrowser, etc'

File: woof-code/rootfs-petbuilds/palemoon/pinstall.sh
Comment (dimkr) [inline woof-code/rootfs-petbuilds/palemoon/pinstall.sh:6]: You'll need to use `chroot . sh -c 'command -v x-www-browser'`, because the shell that runs pinstall.sh runs on the host system and not what you're building (i.e. you're checking if the OS you're running woof-CE on has x-www-browser, not the OS you're building)

File: woof-code/rootfs-petbuilds/palemoon/pinstall.sh
Comment (dimkr) [inline woof-code/rootfs-petbuilds/palemoon/pinstall.sh]: You'll need to wrap this with `chroot`

File: woof-code/rootfs-petbuilds/palemoon/pinstall.sh
Comment (lakshayrohila) [reply woof-code/rootfs-petbuilds/palemoon/pinstall.sh]: Thanks! I'll make the changes.

File: woof-code/rootfs-petbuilds/palemoon/pinstall.sh
Comment (lakshayrohila) [reply woof-code/rootfs-petbuilds/palemoon/pinstall.sh:6]: Please note this line is wrapped within single quotes. So, this is the content of the file `/usr/local/bin/defaultbrowser`, not actually run in the `pinstall.sh` file.

Comment (lakshayrohila) [reply woof-code/rootfs-petbuilds/palemoon/pinstall.sh:6]: Although, I am not really sure if this check is needed. By what I understood by the code you referred to, I thought that `defaultbrowser` always makes this check (in Firefox); so I also added this check here.

Comment (dimkr) [reply woof-code/rootfs-petbuilds/palemoon/pinstall.sh:6]: > not actually run in the pinstall.sh file.

You're right, I admit I reviewed this very quickly while also trying to make a living :)

Comment (dimkr) [reply woof-code/rootfs-petbuilds/palemoon/pinstall.sh:6]: The situation in Puppy is a bit weird because when browsers ask to set themselves as the default they do so through xdg-mime and mimeapps.list: they basically override Puppy's hack to set defaultbrowser.desktop as the default browser. You end up with two default browsers: one that's the default in ROX-Filer and Puppy's replacement for xdg-open, and one for everything else.

Comment (lakshayrohila) [reply woof-code/rootfs-petbuilds/palemoon/pinstall.sh:6]: > one that's the default in ROX-Filer and Puppy's replacement for xdg-open, and one for everything else

I think that although it is not a desired behaviour, still when I use Puppy, I do not mind going to "Set default apps" application and setting the new default (for example if I installed a new browser).

Comment (dimkr) [reply woof-code/rootfs-petbuilds/palemoon/pinstall.sh:6]: The "Set default apps" thing changes defaultbrowser and friends instead of changing mimeapps.list (which sets defaultbrowser as the default for https:// URLs, etc'). In contrast, the "set as default" feature of browsers changes mimeapps.list, while root and spot have separate mimapps.list files.

Comment (dimkr) [reply woof-code/rootfs-petbuilds/palemoon/pinstall.sh:6]: https://github.com/puppylinux-woof-CE/woof-CE/blob/4c8a372cd9bec68d687d2f8018e7121d5563efca/woof-code/3builddistro#L1030

This block in 3builddistro makes Puppy behave pretty much like other distros by 1) removing Puppy's mimeapps.list, which sets defaultbrowser as the handler for https:// URLs, etc' 2) making defaultbrowser run the default handler for https:// URLs instead of the other way around (if the user sets the default browser through the browser or through the file manager, defaultbrowser respects this decision).

This is activated when you build a Puppy without ROX-Filer, because defaultbrowser, Puppy's mimeapps.list and other Puppy-specific hacks exist only to make ROX-Filer happy.

I don't see a clean way to make the browser's "set as default" feature work in a Puppy with ROX-Filer, unless you're willing to make very significant changes like removing Puppy's mimeapps.list and xdg-open.

File: woof-code/rootfs-petbuilds/palemoon/pinstall.sh
Comment (dimkr) [inline woof-code/rootfs-petbuilds/palemoon/pinstall.sh]: Better use `cp -p` or change to `chmod 755 usr/local/bin/defaultbrowser usr/local/bin/defaulthtmlviewer` after the copying

**Review by dimkr — APPROVED** [review]
Looks good to me - again, I have no merge rights

**Comment by lakshayrohila** [comment]
> Looks good to me - again, I have no merge rights

But your reviews are great and helpful! Thanks :+1:!

@peabee, please merge this if you have no objections.

**Comment by peabee** [comment]
> Looks good to me - again, I have no merge rights

@puppylinux-woof-CE/members - please note that this is not a sustainable situation

**Pull request merged**
